odl/space/cu_ntuples.py

```diff
@@ -170,10 +170,22 @@ def element(self, inp=None, data_ptr=None):
             if data_ptr is None:
                 if isinstance(inp, self._vector_impl):
                     return self.element_type(self, inp)
-                elif (isinstance(inp, self.element_type) and
-                      inp.dtype == self.dtype):
-                    return self.element_type(self, inp.data)
+                elif isinstance(inp, self.element_type):
+                    if inp.dtype == self.dtype:
+                        # Simply rewrap for matching dtypes
+                        return self.element_type(self, inp.data)
+                    else:
+                        # Bulk-copy for non-matching dtypes
+                        elem = self.element()
+                        elem[:] = inp
+                        return elem
                 else:
+                    # Array-like input. Need to go through a NumPy array
+                    arr = np.array(inp, copy=False, dtype=self.dtype, ndmin=1)
```

Comment:
Right. But the logic was already there, I just split the behavior for Cuda vectors and other sequence-type things. We need some type of check here, though since we want `CudaRn(2).element(1)` to raise, which wouldn't happen if we just delegated to `__setitem__`. There it makes sense to allow a scalar right hand side.